WcaOnRails/app/controllers/registrations_controller.rb

```diff
@@ -98,6 +108,101 @@ def export
     end
   end
 
+  def import
+    @competition = competition_from_params
+  end
+
+  def do_import
+    competition = competition_from_params
+    file = params[:registrations_import][:registrations_file]
+    required_columns = ["status", "name", "country", "wca id", "birth date", "gender", "email"] + competition.events.map(&:id)
+    # Ensure the CSV file includes all required columns.
+    headers = CSV.read(file.path).first.compact.map(&:downcase)
+    missing_headers = required_columns - headers
+    if missing_headers.any?
+      raise "Missing columns: #{missing_headers.to_sentence}."
+    end
+    registrations = CSV.read(file.path, headers: true, header_converters: :symbol, skip_blanks: true, converters: ->(string) { string&.strip })
+                       .map(&:to_hash)
+                       .reject { |registration| registration.values.all?(&:nil?) }
+                       .select { |registration| registration[:status] == "a" }
+    new_locked_users = []
+    ActiveRecord::Base.transaction do
+      registrations.each do |registration|
+        user = user_for_registration!(registration, new_locked_users)
+        registration_competition_events_attributes = competition.competition_events.map do |competition_event|
+          registration[competition_event.event_id.to_sym] == "1" ? { competition_event_id: competition_event.id } : nil
+        end.compact
+        competition.registrations.create!(
+          user_id: user.id,
+          accepted_at: Time.now,
```

Comment:
Hmm, I think there is more to FIFO checking than just these two timestamps. If registrations get approved once a payment is done, then we would need to know the payment time. On the other hand a lot of platforms don't even have that timestamp, because payments are done via a bank transfer instead of an external provider embedded into one's site. So organizers just approve registrations when the the based on bank transfer dates.

Btw. thinking about this, I've realized that we should probably be checking if the number of approved registrations to be imported doesn't exceed registration limit, right?

Reply:
That's how the mentioned error looks:
![image](https://user-images.githubusercontent.com/17034772/50532070-36e8f900-0b14-11e9-94d6-88506ec3084c.png)

Reply:
Yeah, you're totally right about FIFO information. Let's not worry about this right now, I just wanted to start the discussion.

Checking the registration limit seems like a good idea. Thanks for doing that!